### DISCUSSION TITLE
eProsima Apache license header on all generated code? [13736]

### DISCUSSION THREAD

**Discussion by gavanderhoorn** [eProsima/Fast-DDS-Gen#191] [Ideas]
(apologies if this has been discussed before, I could not find a relevant (closed) issue on the tracker)

It seems files generated by the code generator are all (?) given the following header:

```c
// Copyright 20XX Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

While for open-source projects that would be OK, it seems strange for all generated code to fall under eProsima's copyright, let alone putting all generated code under the Apache license, unconditionally.

Was this done for a particular reason, or is this just a case of "template has this header, template gets copied, final output has this header"?

**Comment by gavanderhoorn** [comment]
Friendly ping?

**Comment by MiguelCompany** [comment]
I would say it is a case of "template has this header, template gets copied, final output has this header". It shouldn't be an issue, since the license lets you use the files on any kind of project.

If it is an issue for you, I would say it is okay if you modify the generated code, but I would like @JaimeMartin to confirm that

**Comment by gavanderhoorn** [comment]
Friendly ping?

**Comment by dch4o** [comment]
Any updates on this issue?

Licensing all the auto-generated header files and source codes feels too harsh to me.

In the case we use fastddsgen to generate message classes and publisher/subscriber with `-example CMake`, publisher and subscriber boilerplate(I would say) examples are generated.
However, these code lines should be customized to use in purpose. Also, there are several virtual functions those can be overriden. Under Apache 2.0 License, we have to point out every customization and clarify all the changes we have made. I think this is not intended and feels quite far away from objectives of Fast-DDS-Gen, which is to provide convenient use of IDL files to generate message types in Fast-DDS.

**Comment by JesusPoderoso** [comment]
Hi @zzodo and @gavanderhoorn, and thanks four you comments.

As @MiguelCompany said, I think that it shouldn't be an issue, since the license lets you use the files on any kind of project.
The Apache License is permissive free software license, and allows to use the software for any purpose, to distribute it, to modify it, and to distribute modified versions of the software under the terms of the license.

For that reason I am moving this issue to the discussion section.

**Reply by gavanderhoorn** [reply]
As @zzodo writes, `Apache-2.0` is permissive, but still requires certain things of consumers of artefacts licensed under it.

> As @MiguelCompany said, I think that it shouldn't be an issue [..]

unfortunately, "I think" and/or "I believe[d]" is legally not very sound :)

Ideally, you'd change the template to not include any license statement, or clarify the situation there.

That would remove legal ambiguity as much as possible.

**Reply by dch4o** [reply]
Thanks @gavanderhoorn for remarking out my point.

And yes, as far as I know Apache 2.0 License requires modification announcement.

@JesusPoderoso Can you confirm this?

**Reply by dch4o** [reply]
@JesusPoderoso 
Friendly ping?

**Reply by JesusPoderoso** [reply]
Hi @gavanderhoorn  and @zzodo .
After internal discussions, we have decided to keep the licences (including Apache license) on the auto-generated code. 

Nonetheless, the generated code is open to modifications. If that does not suite your purposes, you would have two possibilities:
- Using your own Fast DDS-Gen implementation to avoid the licensed auto-generated code, by modifying it in a fork for instance.
- Not use Fast DDS-Gen for type support generation, as it is a supporting tool and it is not required to work with Fast DDS. That means using your own type support implementation.

**Reply by dch4o** [reply]
Hi @JesusPoderoso 

Thanks for your reply. 
Quite surprised to hear that internal decisions have made to keep license in auto-generated codes.

I have one more question to double-check the license modification.

So, is it allowed to fork Fast-DDS-Gen repository and modify auto-generation template (e.g. [src](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src)/[main](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main)/[java](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java)/[com](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java/com)/[eprosima](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java/com/eprosima)/[fastdds](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java/com/eprosima/fastdds)/[idl](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java/com/eprosima/fastdds/idl)/[templates](https://github.com/eProsima/Fast-DDS-Gen/tree/master/src/main/java/com/eprosima/fastdds/idl/templates)
/*.stg) to remove auto-generated licensing notice while keeping the Apache license for the forked repository itself?
For my understanding, your first suggestion makes me feel free to do this. 

Can you confirm this?

Thank you for your time.

**Reply by JesusPoderoso** [reply]
You are right @zzodo , you must keep the license in the header of all files from the repository, but remove the license header generation in each template.stg.

For example, in [this template](https://github.com/eProsima/Fast-DDS-Gen/blob/master/src/main/java/com/eprosima/fastdds/idl/templates/Common.stg), keep the first 13 lines, but remove lines 18 to 30. Hope that helps!

**Reply by dch4o** [reply]
Thank you for clarification!
